sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/data/OpenBisConnector.java

```diff
@@ -0,0 +1,112 @@
+package life.qbic.portal.sampletracking.data;
+
+import ch.ethz.sis.openbis.generic.asapi.v3.IApplicationServerApi;
+import ch.ethz.sis.openbis.generic.asapi.v3.dto.common.search.SearchResult;
+import ch.ethz.sis.openbis.generic.asapi.v3.dto.project.fetchoptions.ProjectFetchOptions;
+import ch.ethz.sis.openbis.generic.asapi.v3.dto.project.search.ProjectSearchCriteria;
+import ch.ethz.sis.openbis.generic.asapi.v3.dto.sample.fetchoptions.SampleFetchOptions;
+import ch.ethz.sis.openbis.generic.asapi.v3.dto.sample.search.SampleSearchCriteria;
+import ch.systemsx.cisd.common.spring.HttpInvokerUtils;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import life.qbic.datamodel.dtos.portal.PortalUser;
+import life.qbic.portal.sampletracking.view.projects.viewmodel.Project;
+import life.qbic.portal.sampletracking.view.samples.viewmodel.Sample;
+
+/**
+ * <b>short description</b>
+ *
+ * <p>detailed description</p>
+ *
+ * @since <version tag>
+ */
+public class OpenBisConnector implements ProjectRepository, SampleRepository {
+
+  private final String sessionToken;
+
+  private final IApplicationServerApi api;
+
+  private static final int TIMEOUT = 10_000;
+
+  private final List<ch.ethz.sis.openbis.generic.asapi.v3.dto.project.Project> cachedProjects = new ArrayList<>();
+  private final Map<String, List<ch.ethz.sis.openbis.generic.asapi.v3.dto.sample.Sample>> cachedSamples = new HashMap<>();
+
+
+
+  public OpenBisConnector(Credentials credentials, PortalUser portalUser,
+      String openBisUrl) {
+    this.api = HttpInvokerUtils.createServiceStub(IApplicationServerApi.class,
+        openBisUrl + "/rmi-application-server-v3", TIMEOUT);
+    this.sessionToken = api.loginAs(credentials.getUser(), credentials.getPassword(),
+        portalUser.getAuthProviderId());
+  }
+
+  @Override
+  public List<Project> findAllProjects() {
+    if (cachedProjects.isEmpty()) {
+      ProjectFetchOptions fetchOptions = new ProjectFetchOptions();
+      fetchOptions.withSpace();
+      SearchResult<ch.ethz.sis.openbis.generic.asapi.v3.dto.project.Project> projectSearchResult =
+          api.searchProjects(sessionToken, new ProjectSearchCriteria(), fetchOptions);
+
+      List<ch.ethz.sis.openbis.generic.asapi.v3.dto.project.Project> projects = projectSearchResult.getObjects().stream()
+          .filter(hasValidProjectCode()).sorted(Comparator.comparing(
+              ch.ethz.sis.openbis.generic.asapi.v3.dto.project.Project::getRegistrationDate).reversed())
```

Comment:
NIT: is registration date useful here? maybe modification date or even code would be more intuitive

Reply:
Good question. Maybe modification date would be better